### PR TITLE
Allow the collector to consume data when node is configured with tls.

### PR DESCRIPTION
**Pull request by EmilLa1**
## Purpose
Fixes https://github.com/Concordium/concordium-node/issues/989

## Changes

If the `grpc_host` (the node) is configured for using TLS, then the collector now sets up a tls connection based on the system trust store.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: collector/src/bin/collector.rs [hunk 1]
```diff
@@ -196,7 +196,18 @@ async fn collect_data<'a>(
...
+    // Set up a tls client if required using the locally installed trust store.
+    let endpoint = Channel::from_shared(grpc_host)?.timeout(Duration::from_secs(grpc_timeout));
+    let channel = if endpoint.uri().scheme().map_or(false, |x| x == &http::uri::Scheme::HTTPS) {
+        endpoint.tls_config(ClientTlsConfig::new())?
+    } else {
+        endpoint
+    };
+
+    //    let channel =
+    // Channel::from_shared(grpc_host)?.timeout(Duration::from_secs(grpc_timeout));
```
Comment (abizjak) [inline collector/src/bin/collector.rs:209]: What is this?

File: collector/src/bin/collector.rs [hunk 2]
```diff
@@ -196,7 +196,18 @@ async fn collect_data<'a>(
 
     // Setting up Client
     let grpc_timeout = conf.grpc_timeout;
-    let channel = Channel::from_shared(grpc_host)?.timeout(Duration::from_secs(grpc_timeout));
+
+    // Set up a tls client if required using the locally installed trust store.
+    let endpoint = Channel::from_shared(grpc_host)?.timeout(Duration::from_secs(grpc_timeout));
+    let channel = if endpoint.uri().scheme().map_or(false, |x| x == &http::uri::Scheme::HTTPS) {
+        endpoint.tls_config(ClientTlsConfig::new())?
+    } else {
```
Comment (abizjak) [inline collector/src/bin/collector.rs:204]: I am not sure why this works. By default no TLS roots are enabled. tonic has a feature https://docs.rs/tonic/latest/tonic/#feature-flags `tls-roots` which is not enabled by default.

Perhaps it's enabled by another dependency? But we should be explicit about this regardless.

File: collector/src/bin/collector.rs (hunk 1 above)
Comment (EmilLa1) [reply collector/src/bin/collector.rs]: No idea. It's gone now :)

File: collector/src/bin/collector.rs (hunk 2 above)
Comment (EmilLa1) [reply collector/src/bin/collector.rs:204]: Explicit now.

File: collector/src/bin/collector.rs [hunk 3]
```diff
@@ -196,7 +196,15 @@ async fn collect_data<'a>(
 
     // Setting up Client
     let grpc_timeout = conf.grpc_timeout;
-    let channel = Channel::from_shared(grpc_host)?.timeout(Duration::from_secs(grpc_timeout));
+
+    // Set up a tls client if required using the locally installed trust store.
+    let endpoint = Channel::from_shared(grpc_host)?.timeout(Duration::from_secs(grpc_timeout));
+    let channel = if endpoint.uri().scheme().map_or(false, |x| x == &http::uri::Scheme::HTTPS) {
```
Comment (td202) [inline collector/src/bin/collector.rs:202]:
```suggestion
    let channel = if endpoint.uri().scheme() == Some(&http::uri::Scheme::HTTPS) {
```

File: collector/README.md [hunk 4]
```diff
@@ -9,3 +9,6 @@ Assuming you have a node running locally with GRPC V2 available at `127.0.0.1:20
 ```console
 $> cargo run -- --collector-url http://127.0.0.1:8080/post/nodes --grpc-host http://127.0.0.1:20000  --node-name "my-node-name"
 ```
+
+### TLS
+If the node is configured to be using TLS connections then one has to configure the collector (the `--grpc-host` flag) to be using the URL of the domain of which the certificate was issued for.
```
Comment (td202) [inline collector/README.md:14]:
```suggestion
If the node is configured to use TLS connections then the collector must be configured (with the `--grpc-host` flag) to use the URL with the domain for which the certificate was issued.
```

File: collector/src/bin/collector.rs (hunk 3 above)
Comment (abizjak) [reply collector/src/bin/collector.rs:202]: So this is resolved, but the code is unchanged. What is the meaning of this?

**Review by abizjak — APPROVED** [review]
The code looks OK assuming Thomas' comment is addressed.

I assume you tested this right? Just to make sure. You can easily test it on the service node on stagenet which is configured with TLS.

File: collector/src/bin/collector.rs (hunk 3 above)
Comment (EmilLa1) [reply collector/src/bin/collector.rs]: Probably clicked the wrong button by accident

**Comment by EmilLa1** [comment]
> The code looks OK assuming Thomas' comment is addressed.
> 
> I assume you tested this right? Just to make sure. You can easily test it on the service node on stagenet which is configured with TLS.

Yes, I checked that it works with and without TLS.

**Review by limemloh — APPROVED** [review]

**Review by td202 — APPROVED** [review]

**Pull request merged**